### PR TITLE
chore(githubrelease) factorize githubrelease plugin out of the github package

### PR DESCRIPTION
**Pull request by dduportal**
This PR is a refactor of the github plugin to allow continuing the work in #412.
No functionnal changes (hence the chore) label.

It introduces the following changes:

- Move all the "githubrelease" logic into a separated plugin packaged as `githubrelease` with its own unit tests
- Introduces news interfaces in the `github` plugin to define the expect contract of this plugin, and allowing to mock it out on other elements
- Refactor the releases, tags, label and changelog GH v4 API logics into their own custom query types to allow mocking
- Moved github tests to unit tests (with mocking the GH API)
- Cleanup methods and attributes on both plugins (using private element as much as possible)

## Test

To test this pull request, you can run the following commands:

```shell
go build -o ./dist/updatecli && ./dist/updatecli diff --config examples/updatecli.generic/githubRelease.yaml
```

### CODE REVIEW

File: pkg/core/pipeline/source/main.go [hunk 1]
```diff
@@ -179,22 +184,18 @@ func (s *Source) Unmarshal() (sourcer Sourcer, changelog Changelog, err error) {
...
-		err := mapstructure.Decode(s.Config.Spec, &githubSpec)
-
-		if err != nil {
+		if err := mapstructure.Decode(s.Config.Spec, &sourceSpec); err != nil {
 			return nil, nil, err
 		}
 
-		g, err := github.New(githubSpec)
-
+		grPrResource, err := githubrelease.New(sourceSpec)
```
Comment (olblak) [inline pkg/core/pipeline/source/main.go:193]: What's the meaning of the variable grPrResource?

File: pkg/plugins/github/changelog.go [hunk 2]
```diff
@@ -17,70 +18,66 @@ type Changelog struct {
...
+}
+type queriedRelease struct {
+	Description string
+	Url         string
+	PublishedAt time.Time
+}
 
-	client := g.NewClient()
+// Changelog returns a changelog description based on a release name
+func (g *Github) Changelog(version version.Version) (string, error) {
```
Comment (olblak) [inline pkg/plugins/github/changelog.go:52]: Without being a blocker, for now, I think it wouldn't have been better to create a new struct directly in the changelog package such as "changelogInformation or "changelogVersion"?
We now have an unnecessary strong binding between this package and version.

File: pkg/plugins/github/label.go [hunk 3]
```diff
@@ -7,76 +7,77 @@ import (
...
-					Node   struct {
-						ID          string
-						Name        string
-						Description string
-					}
-				}
-			} `graphql:"labels(last: 5, before: $before)"`
-		} `graphql:"repository(owner: $owner, name: $repository)"`
-	}
+	var query labelsQuery
```
Comment (olblak) [inline pkg/plugins/github/label.go:77]: I really like this refactoring :)

**Review by olblak — DISMISSED** [review]
What a nice refactoring.
Excepted minor questions, this pull request looks really great

**Comment by olblak** [comment]
@dduportal  I propose to trigger a release just with the current change

File: pkg/core/pipeline/source/main.go (hunk 1 above)
Comment (dduportal) [reply pkg/core/pipeline/source/main.go:193]: Good catch, it is a typo. Should be the following:

```suggestion
		ghrResource, err := githubrelease.New(sourceSpec)
```

WDYT?

File: pkg/plugins/github/changelog.go (hunk 2 above)
Comment (dduportal) [reply pkg/plugins/github/changelog.go:52]: Not sure to understand?

- Using a private struct for a public method is not possible (how would the externe caller be able to use the struct)?
- The intent is to change the signature of the method `Changelog` globally (see the changes on the `Changelog` interface below) and use the version.Version as argument type. Rationale is that it depends on the implementation of the changelog if either the parsed or originl version are required.

**Comment by dduportal** [comment]
> @dduportal I propose to trigger a release just with the current change

No strong advise so go for it if you feel it

File: pkg/core/pipeline/source/main.go (hunk 1 above)
Comment (dduportal) [reply pkg/core/pipeline/source/main.go]: I went for `newGhrResource`. WDYT?

File: pkg/plugins/github/changelog.go (hunk 2 above)
Comment (olblak) [reply pkg/plugins/github/changelog.go:52]: I was thinking to create a public struct defined in the changelog package so we wouldn't need to import the package version. But the way you implemented it is very good

**Review by olblak — APPROVED** [review]

**Pull request merged**